Experiment: Put directives on <figure> element and insert lightbox as its child

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This revises the HTML structure for the experimental [image lightbox](https://github.com/WordPress/gutenberg/pull/50373) by ensuring there is no wrapping `<div>`.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Addresses #51068 and related issues as detailed in the [following comment](https://github.com/WordPress/gutenberg/issues/51068#issuecomment-1568033497), namely that containing a wrapping `<div>` causes existing styles and functionality to break.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds directives to the `<figure>` element using the tag processor, and does a regex for the `<figure>` closing tag, adding the lightbox HTML as its child.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Enable Gutenberg > Experiments > Core Blocks
2. Add an image to a post
3. Enable Advanced > Behaviors > Lightbox in the image block settings
4. Publish the post and inspect the source — see that the `data-wp-context` and `data-wp-island` attributes have been added to the `<figure>` element.
5. Click on the image; the lighbox should work.


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

artemiomorales committed:
Put directives on <figure> element and insert lightbox as its child

artemiomorales committed:
Fix PHP spacing

SantosGuillamot reviewed (approved):
I've tested it, and it seems to work fine. Thanks for the pull request 🙂 

It seems that there is a test not passing though.

gziolo commented:
> Addresses https://github.com/WordPress/gutenberg/issues/51068 and related issues as detailed in the https://github.com/WordPress/gutenberg/issues/51068#issuecomment-1568033497, namely that containing a wrapping <div> causes existing styles and functionality to break.

I tested a few scenarios, and it resolved the listed issues.

artemiomorales commented:
@gziolo Who else should we ping to move this fix forward?

artemiomorales committed:
Ensure <figure> directives are only rendered in body content

The <figure> directives were rendering not only in the body
content, but also the lightbox. This ensures the directives
do not appear in the lightbox.

artemiomorales commented:
I realized the directives were being unnecessarily duplicated in the `<figure>` in the lightbox; that has now been fixed.

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -59,48 +59,54 @@ function render_block_core_image( $attributes, $content ) {
+			array(
+				'tag_name' => 'figure',
+			)
+		);
```
note that it's fine to use `$w->next_tag( 'figure' )` here too

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -59,48 +59,54 @@ function render_block_core_image( $attributes, $content ) {
+			)
+		);
+		$w->add_class( 'wp-lightbox-container' );
+		$w->set_attribute( 'data-wp-island', '' );
```
this is going to produce `data-wp-island=""` but if all we need is the attribute as a boolean we can pass `set_attribute( 'data-wp-island', true )` and it will generate `data-wp-island`

in JS and in the browser both should be equivalent

luisherranz replied:
Funny thing, the browser works the other way around:

- `element.setAttribute("data-wp-island", "")` -> `data-wp-island`
- `element.setAttribute("data-wp-island", "true")` -> `data-wp-island="true"`

luisherranz replied:
But FWIW, I think `true` makes more sense. That's the way Preact/React work, so that's what we did for the Interactivity API. I'm happy that the HTML Tag Processor does the same thing 😄

We have an exception tho ([copied from Preact](https://github.com/WordPress/block-interactivity-experiments/pull/236)): `data-` and `aria-` attributes. Those should be enumerated. If you pass `true` or `false`, instead of adding/removing the element, it writes it as a string. It makes sense when working with things like `aria-expanded`, where you need the attribute to be either `aria-expanded="true"` and `aria-expanded="false"` instead of appear/disappear. Forcing people to remember to make a selector just to turn `true`/`false` into `"true"`/`"false"` could cause accessibility problems.

dmsnell replied:
@luisherranz it's of note that `"true"` is a different value than `""`, which is why the Tag Processor doesn't output `"true"`. it's really complicated.

for `aria-expanded` this is fine because the spec wants a string value of `"true"`, but if something is looking for a boolean HTML attribute and reads `"false"` it will consider it true 🙃. an example of this is setting `disabled="false"`, which disables the element it's on.

`true` is added in the Tag Processor so that it's possible to differentiate an attribute with an empty string value and an attribute with no value. similarly if you pass `false` as the value it will remove the attribute entirely, which is the only way to do things like set `disabled` to `false`.

just wanted to share; the code works as-is.

dmsnell replied:
sidenote: part of why it's confusing is that all HTML attribute values are strings, so the only way to communicate a boolean value is by the presence or absence of any value.

in addition, if the attribute _has_ a value, then it must be exactly the empty string `""` or exactly a case-insensitive match for the attribute's canonical name, like `"disabled"`

> Forcing people to remember to make a selector just to turn true/false into "true"/"false" could cause accessibility problems.

I agree, but that's kind of my point as well. this is water under the bridge because different HTML attributes demand that we know the difference and use the right one; that's not caused by Preact or the Tag Processor - those are just trying their best to conform to what is required so that they don't unintentionally flip a boolean value.

luisherranz replied:
> if all we need is the attribute as a boolean we can pass `set_attribute( 'data-wp-island', true )` and it will generate `data-wp-island` [...] in JS and **in the browser both should be equivalent**

What I was trying to say is that that's not how the browser works. In the browser, `elem.setAttribute("data-wp-island", true)` produces `data-wp-island="true"`, not `data-wp-island`. To produce `data-wp-island` you need to use `elem.setAttribute("data-wp-island", "")`.

dmsnell replied:
we're _almost_ saying the same thing I think 😄 

_but_! be careful with the JS in the browser and using `setAttribute( …, false )` because it's not doing what the Tag Processor is doing. its behavior follows type casting, evident when setting the value to `null` (the `"false"` you get is the result of casting `false` to a string, not the result of setting _false_ as the value)

if you call `element.setAttribute( 'disabled', false )` you will accidentally disable the element you meant to enable. in those cases the only way to _un_-disable the element is to call `element.removeAttrbiute( 'disabled' )`

so this is the inherent danger that I was referring to, because JS and the DOM API is the source of the problem.

luisherranz replied:
> because JS and the DOM API is the source of the problem

That's the reason [I was happy](https://github.com/WordPress/gutenberg/pull/51089#discussion_r1211939558) about the Tag Processor API 😄

ajlende commented on packages/block-library/src/image/index.php:
```diff
@@ -59,48 +59,54 @@ function render_block_core_image( $attributes, $content ) {
+			</div>
 HTML;
+
+		return preg_replace( '/<\/figure>/', $lightbox_html . '</figure>', $body_content );
```
If you're not using any regex features, it's better to use `str_replace`

glendaviesnz commented:
This tests well for me as a fix for the Gallery layout issue.

artemiomorales committed:
Improve syntax

SantosGuillamot committed:
Change directives suffix and `data-wp-island`
